samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -258,7 +290,9 @@ public void onBarrierCreated(String version) {
 
     public void onBarrierStateChanged(final String version, ZkBarrierForVersionUpgrade.State state) {
       LOG.info("JobModel version " + version + " obtained consensus successfully!");
+      metrics.barrierStateChange.inc();
       if (ZkBarrierForVersionUpgrade.State.DONE.equals(state)) {
+        metrics.singleBarrierRebalancingTime.update(System.nanoTime() - startTime);
```

Comment:
I think we should update the timer metric even in case of a time out state change. That way, we can correlate the metric value with the state change - if the difference in time == barrier_timeout-ms, then the barrier has timed out. What do you think? 

I also want to point out a potential issue here: `startTime` is initialized in `doProcessorChange` which is only invoked in the leader JC. For all other processors, it will be 0 and result high and incorrect timer value. Perhaps it isn't obvious from the code. But the leader also receives barrierCreation notification. So, you can create the `startTime` within the handler and initialize it in `ZkBarrierListenerImpl#onBarrierCreated`. Does this sound reasonable to you?